### PR TITLE
feat: process documents directly on Backend and passing task id to Langflow

### LINKED ISSUE #919: [Feature]: Better ingestion exploration can we upload docs directly to docling and if langflow can use that refernce?
### Feature Area

Ingestion (document processing, upload, Docling)

### Problem Description

Currently file is hoping from OpenRAG BE> LF BE> Docling Serve

### Proposed Solution

Currently file is hoping from OpenRAG BE>  Docling Serve

### Use Case

Better for deployemnt and ingestion speeds.
Also Optimised the network costs.

### Alternatives Considered

_No response_

### Priority

Nice to have

### Additional Context

_No response_

### Contribution

- [x] I would be willing to help implement this feature.
- [x] I can help test this feature once implemented.

### Checklist

- [x] I have searched existing issues and discussions to ensure this feature hasn't been requested before.

### PR DESCRIPTION
**Pull request by lucaseduoli**
This pull request updates the Docling integration in the ingestion flow to add support for polling existing Docling Serve tasks by task ID, and improves compatibility and maintainability. The changes primarily add a new `task_id` parameter to the Docling component, update environment variable handling, and adjust type names for consistency.

**Docling Serve component improvements:**

* Added support for an optional `task_id` parameter to the Docling Serve component, allowing users to poll for results of previously submitted tasks instead of uploading new files. The component now checks for `task_id` and, if present, skips file upload and polls for the task's result. [[1]](diffhunk://#diff-72350f42472113fe21ccf03bcd014852fe367b5907ecf9e347f8a7c852037d0dR1331) [[2]](diffhunk://#diff-72350f42472113fe21ccf03bcd014852fe367b5907ecf9e347f8a7c852037d0dL1450-L1460)
* Refactored the component code to include a `_poll_and_fetch_result` helper method for polling and result retrieval, and updated `load_files_base` and `process_files` methods to handle the new logic.

**Type and compatibility updates:**

* Changed the output type and base class of the Docling Serve node from `DataFrame` to `Table` for better compatibility and clarity. [[1]](diffhunk://#diff-72350f42472113fe21ccf03bcd014852fe367b5907ecf9e347f8a7c852037d0dL13-R13) [[2]](diffhunk://#diff-72350f42472113fe21ccf03bcd014852fe367b5907ecf9e347f8a7c852037d0dL1313-R1321) [[3]](diffhunk://#diff-72350f42472113fe21ccf03bcd014852fe367b5907ecf9e347f8a7c852037d0dL1378-R1382) [[4]](diffhunk://#diff-72350f42472113fe21ccf03bcd014852fe367b5907ecf9e347f8a7c852037d0dL28-R31)
* Updated dependencies for `docling_core`, `pydantic`, and added a pinned version for `lfx` to ensure compatibility.

**Configuration and environment variable updates:**

* Added `DOCLING_TASK_ID` as a new environment variable in `docker-compose.yml` and included it in `LANGFLOW_VARIABLES_TO_GET_FROM_ENVIRONMENT` so it can be passed into the container and used by the application.

**Other minor updates:**

* Updated metadata such as `last_updated`, code hash, and frontend node folder ID to reflect the new changes. [[1]](diffhunk://#diff-72350f42472113fe21ccf03bcd014852fe367b5907ecf9e347f8a7c852037d0dL831-R831) [[2]](diffhunk://#diff-72350f42472113fe21ccf03bcd014852fe367b5907ecf9e347f8a7c852037d0dL881-R881) [[3]](diffhunk://#diff-72350f42472113fe21ccf03bcd014852fe367b5907ecf9e347f8a7c852037d0dL1340-R1341)

These updates make the Docling integration more flexible and robust, allowing for asynchronous workflows and improved environment management.

Closes #919

## Summary by CodeRabbit

* **New Features**
  * Integrated Docling document processing for async file conversion and combined upload+ingest flow.
  * Added multi-model embedding support and expanded embedding model options.
  * Improved table-structured data handling in ingestion flows.
  * Exposed Docling configuration/env for deployment and runtime.

* **Bug Fixes**
  * Improved error handling, polling/timeouts, and clearer ingestion failure messages.
  * Better timing and diagnostic logging for ingestion operations.

* **Chores**
  * Removed the delete-after-ingest option from ingestion endpoints.

* **Tests**
  * Added comprehensive unit tests for Docling integration and polling logic.

### CODE REVIEW

File: src/connectors/langflow_connector_service.py [hunk 1]
```diff
@@ -19,13 +19,17 @@ def __init__(
         self,
         task_service=None,
         session_manager=None,
+        flows_service=None,
+        docling_service=None,
```
Comment (mpawlow) [inline src/connectors/langflow_connector_service.py:23]: ### (a) [Blocker] `LangflowConnectorService` Missing `self.docling_service` Attribute Causes `AttributeError`

#### Problem

- `LangflowConnectorService.__init__` accepts a `docling_service` parameter and forwards it to `LangflowFileService`, but **never stores it as an instance attribute** (`self.docling_service = docling_service` is absent)
- `LangflowConnectorFileProcessor.__init__` reads `langflow_connector_service.docling_service` at line 625 of `processors.py`
- Since the attribute does not exist on `LangflowConnectorService`, every invocation of `LangflowConnectorFileProcessor` raises `AttributeError: 'LangflowConnectorService' object has no attribute 'docling_service'`
- This path is exercised whenever a Langflow-backed connector syncs files (e.g., Google Drive, OneDrive, SharePoint via `process_files_from_connection` or `sync_specific_files`)

#### Code References

- `src/models/processors.py` — line 625 (`docling_service=langflow_connector_service.docling_service`)
- `src/connectors/langflow_connector_service.py` — lines 266, 390 (instantiation sites for `LangflowConnectorFileProcessor`)

#### Potential Solution

Add the missing assignment inside `LangflowConnectorService.__init__`:

```python
def __init__(
    self,
    task_service=None,
    session_manager=None,
    flows_service=None,
    docling_service=None,
):
    self.task_service = task_service
    self.session_manager = session_manager
    self.docling_service = docling_service           # <-- add this line
    self.connection_manager = ConnectionManager()
    self.langflow_service = LangflowFileService(
        flows_service=flows_service, docling_service=docling_service
    )
```

File: src/models/processors.py [hunk 2]
```diff
@@ -261,9 +262,7 @@ async def process_document_standard(
             if original_filename:
                 slim_doc["filename"] = original_filename
         else:
-            from utils.docling_client import convert_file
-
-            full_doc = await convert_file(file_path, httpx_client=clients.docling_http_client)
+            full_doc = await self.docling_service.convert_file(file_path)
```
Comment (mpawlow) [inline src/models/processors.py:265]: ### (b) [Critical] `S3FileProcessor` Does Not Pass `docling_service` to Base Class

#### Problem

- `S3FileProcessor.__init__` calls `super().__init__(document_service, models_service)` (line 736 of `processors.py`) without the `docling_service` keyword argument
- `TaskProcessor.__init__` signature is `def __init__(self, document_service=None, models_service=None, docling_service=None)`, so `self.docling_service` is silently set to `None`
- `process_document_standard` (called from `S3FileProcessor.process_item`) invokes `self.docling_service.convert_file(file_path)` for all non-`.txt`/`.md` files — this will raise `AttributeError: 'NoneType' object has no attribute 'convert_file'`
- S3 document ingestion (PDF, DOCX, etc.) is entirely broken as a result

#### Potential Solution

Update `S3FileProcessor.__init__` to include `docling_service`:

```python
super().__init__(
    document_service,
    models_service,
    docling_service=document_service.docling_service if document_service else None,
)
```

File: src/services/langflow_file_service.py [hunk 3]
```diff
@@ -517,43 +523,42 @@ async def upload_and_ingest_file(
...
-        logger.debug("[LF] Starting combined upload and ingest operation")
+        logger.debug("[LF] Starting Docling-based upload and ingest operation")
 
-        # Step 1: Upload the file
+        filename, content, _ = file_tuple
+
+        # Step 1: Upload the file to Docling
         try:
-            upload_result = await self.upload_user_file(file_tuple, jwt_token=jwt_token)
+            task_id = await self.docling_service.upload_to_docling_direct_async(filename, content)
```
Comment (mpawlow) [inline src/services/langflow_file_service.py]: ### (c) [Major] No Guard Against `docling_service=None` in `upload_and_ingest_file`

#### Problem

- `LangflowFileService.upload_and_ingest_file` calls `self.docling_service.upload_to_docling_direct_async(...)` (line 547) with no `None` check
- If `docling_service` is not injected (e.g., Docling is not deployed, or a code path constructs `LangflowFileService` without passing it), the call raises an opaque `AttributeError: 'NoneType' object has no attribute 'upload_to_docling_direct_async'` instead of a user-friendly error
- The `DISABLE_INGEST_WITH_LANGFLOW` feature flag controls which code path is taken, so mixed-mode deployments could reach this path with `docling_service=None`

#### Potential Solution

Add an explicit guard at the top of the method:

```python
if self.docling_service is None:
    raise RuntimeError(
        "DoclingService is not configured. Ensure DOCLING_SERVE_URL is set "
        "and the service was injected correctly."
    )
```

#### Alternative Solutions

- Make `docling_service` a required constructor argument on `LangflowFileService` and validate it at construction time rather than at call time.

File: src/services/langflow_file_service.py [hunk 4]
```diff
@@ -563,10 +568,11 @@ async def upload_and_ingest_file(
                 extra={"settings": settings, "tweaks": final_tweaks},
             )
 
-        # Step 3: Run ingestion
+        # Step 3: Run ingestion via Langflow
```
Comment (mpawlow) [inline src/services/langflow_file_service.py:604]: ### (d) [Major] Docling Task Orphaned When Langflow Ingestion Fails

#### Problem

- `upload_and_ingest_file` follows a two-step flow: (1) upload to Docling → (2) run Langflow ingestion
- If Langflow ingestion fails (step 2), the Docling task created in step 1 is never cleaned up
- Docling-serve retains the task in memory/storage, leaking resources for every failed ingestion attempt
- For long-running or high-volume workflows this can exhaust Docling-serve's task store or memory

#### Code References

- `src/services/langflow_file_service.py` — lines 545–558 (Docling upload step, stores `task_id`)
- `src/services/langflow_file_service.py` — lines 571–593 (Langflow ingestion step, re-raises on failure with no cleanup)

#### Potential Solution

Add a `try/finally` block (or an `except` clause) around the ingestion step that issues a `DELETE /v1/tasks/{task_id}` request to Docling-serve on failure:

```python
try:
    ingest_result = await self.run_ingestion_flow(...)
except Exception:
    try:
        await self.docling_service.cancel_task(task_id)
    except Exception as cleanup_err:
        logger.warning("[LF] Failed to cancel Docling task after ingestion error",
                       task_id=task_id, error=str(cleanup_err))
    raise
```

File: src/services/docling_service.py [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+
+    async def upload_to_docling_direct_async(self, filename: str, file_content: bytes) -> str:
+        """
+        Upload a file to Docling Serve asynchronously using direct multipart/form-data upload.
+        """
+        options = self._build_docling_options()
+        
+        # Docling serve async multipart endpoint /v1/convert/file/async
+        # Options are passed as form data
+        data = {k: str(v) if isinstance(v, bool) else v for k, v in options.items() if not isinstance(v, dict)}
```
Comment (mpawlow) [inline src/services/docling_service.py]: ### (e) [Major] Boolean Form Data Values Sent as Python-Capitalized `"True"`/`"False"`

#### Problem

- `upload_to_docling_direct_async` serializes options for the multipart form payload with `str(v) if isinstance(v, bool) else v` (line 95)
- `str(True)` in Python produces `"True"` and `str(False)` produces `"False"` — these are not standard HTTP form boolean representations
- Web frameworks (e.g., FastAPI's form parser) typically accept `"true"`, `"false"`, `"1"`, `"0"` as booleans; `"True"` is not universally recognized
- If Docling-serve rejects or silently ignores these values, OCR and table-structure options would be silently disabled regardless of user configuration

#### Potential Solution

Use lowercase string conversion for boolean values to match standard HTTP form conventions:

```python
data = {
    k: str(v).lower() if isinstance(v, bool) else v
    for k, v in options.items()
    if not isinstance(v, dict)
}
```

File: src/services/langflow_file_service.py
Comment (mpawlow) [inline src/services/langflow_file_service.py]: ### (f) [Normal] `delete_after_ingest` Parameter Silently Ignored Without Deprecation

#### Problem

- `upload_and_ingest_file` retains the `delete_after_ingest: bool = True` parameter in its signature (line 519), but the parameter is never read or acted upon anywhere in the method body
- The docstring acknowledges this ("ignored as we don't upload to Langflow"), but callers — including `LangflowFileProcessor.process_item` (line 916 of `processors.py`) — still pass `delete_after_ingest=self.delete_after_ingest`, expecting it to have an effect
- This creates a false API contract: callers are told the parameter works when it does not

#### Code References

- `src/models/processors.py` — line 916 (`delete_after_ingest=self.delete_after_ingest`)

#### Potential Solution

Remove the `delete_after_ingest` parameter entirely from the signature since it no longer applies (files go directly to Docling, never to Langflow):

```python
async def upload_and_ingest_file(
    self,
    file_tuple,
    session_id: Optional[str] = None,
    tweaks: Optional[Dict[str, Any]] = None,
    settings: Optional[Dict[str, Any]] = None,
    jwt_token: Optional[str] = None,
    owner: Optional[str] = None,
    ...
) -> Dict[str, Any]:
```

Update all call sites accordingly and remove the `delete_after_ingest` field from `LangflowFileProcessor`.

#### Alternative Solutions

- If backwards compatibility must be maintained temporarily, decorate with `@deprecated` or emit a `DeprecationWarning` when the parameter is passed with a non-default value.

File: src/services/docling_service.py [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+    async def get_docling_result_async(
+        self,
+        task_id: str,
+        poll_interval: float = 1.0,
+        timeout: float = 600.0
+    ) -> Dict[str, Any]:
+        """
+        Poll Docling Serve for the result of an async conversion task.
+        """
+        elapsed = 0.0
```
Comment (mpawlow) [inline src/services/docling_service.py]: ### (g) [Normal] Dead Variable `elapsed = 0.0` in `get_docling_result_async`

#### Problem

- `get_docling_result_async` declares `elapsed = 0.0` at line 137, but the variable is never read or modified within that method
- The actual elapsed tracking is done inside `_poll_result` via its own independent `elapsed` local variable (line 153)
- Dead variables add noise and can mislead readers into thinking timeout tracking is happening at the wrong level

#### Potential Solution

Remove the dead variable from `get_docling_result_async`:

```python
async def get_docling_result_async(
    self,
    task_id: str,
    poll_interval: float = 1.0,
    timeout: float = 600.0
) -> Dict[str, Any]:
    """Poll Docling Serve for the result of an async conversion task."""
    client = self._get_client()
    should_close = client != self.httpx_client
    ...
```

File: src/services/docling_service.py [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+        """
+        if docling_url:
+            self.docling_url = docling_url.rstrip("/")
+        else:
+            from api.docling import DOCLING_SERVICE_URL
+            self.docling_url = DOCLING_SERVICE_URL
+            
+        self.httpx_client = httpx_client
+
+    def _get_client(self) -> httpx.AsyncClient:
```
Comment (mpawlow) [inline src/services/docling_service.py:63]: ### (h) [Normal] `_get_client()` Creates a New `AsyncClient` on Every Call — No Connection Pooling

#### Problem

- `_get_client()` returns either the injected `self.httpx_client` or creates a fresh `httpx.AsyncClient(...)` every call (line 67)
- A single `convert_file` call invokes `upload_to_docling_direct_async` and `get_docling_result_async` — each calls `_get_client()` independently, creating two separate short-lived clients with no shared connection pool
- High-throughput ingestion scenarios will see excessive TCP connection churn and increased latency for TLS handshakes
- Additionally, the polling loop in `_poll_result` reuses a client only within a single poll session, but a second independent call to `get_docling_result_async` would create yet another client

#### Code References

- `src/services/docling_service.py` — lines 103–104, 138–139 (call sites)

#### Potential Solution

Create a module-level or class-level default client that is reused when no `httpx_client` is injected:

```python
class DoclingService:
    _default_client: Optional[httpx.AsyncClient] = None

    def _get_client(self) -> httpx.AsyncClient:
        if self.httpx_client:
            return self.httpx_client
        if DoclingService._default_client is None or DoclingService._default_client.is_closed:
            DoclingService._default_client = httpx.AsyncClient(
                timeout=httpx.Timeout(300.0, connect=10.0)
            )
        return DoclingService._default_client
```

#### Alternative Solutions

- Cache the fallback client as `self._local_client` on the instance, initialized once in `__init__` when `httpx_client` is not provided.

File: src/config/settings.py [hunk 8]
```diff
@@ -600,6 +601,16 @@ async def refresh_patched_client(self):
             finally:
                 self._patched_async_client = None
 
+    @property
```
Comment (mpawlow) [inline src/config/settings.py:623]: ### (i) [Normal] Lazy `docling_service` Property in `OpenRAGConfig` Is Not Thread-Safe

#### Problem

- The `docling_service` property in `src/config/settings.py` (lines 604–612) lazily initializes `_docling_service` with a simple `if self._docling_service is not None` guard, but takes no lock
- Under concurrent async requests, if two coroutines both observe `_docling_service is None` before either completes initialization, two `DoclingService` instances are created; one is silently discarded
- While asyncio is single-threaded in theory, `asyncio.create_task` and thread-executor calls mean the property can be called from multiple execution contexts before the first call completes; the real risk is across threads when using `ThreadPoolExecutor`

#### Code References

- `src/config/settings.py` — line 327 (`self._docling_service = None` in `__init__`)

#### Potential Solution

- Initialize `_docling_service` eagerly alongside `docling_http_client` in `main.py`'s `lifespan` block (where all other services are wired), rather than lazily on first property access. This matches the project's established initialization pattern.

#### Alternative Solutions

- Use `threading.Lock` around the lazy initialization, consistent with how `self._client_init_lock` is already used for the patched async client elsewhere in the class.

File: src/services/docling_service.py [hunk 9]
```diff
@@ -0,0 +1,198 @@
+import base64
```
Comment (mpawlow) [inline src/services/docling_service.py]: ### (j) [Minor] No Unit Tests for the New `DoclingService` Class

- Note: This is Minor severity. Please feel free to ignore or optionally implement
   - Probably a task for post 1.0 release

#### Problem

- `DoclingService` introduces non-trivial async logic: async HTTP uploads, a polling loop with exponential progression, timeout handling, error propagation from `DoclingServeError`, and boolean-to-string serialization
- None of this logic is covered by tests
- The existing test suite (`tests/unit/`, `tests/integration/`) has no additions for this file
- Critical edge cases — Docling returning `"failure"` status, the timeout path, `json_content` missing from the result, connection errors mid-poll — are completely untested

#### Potential Solution

Add `tests/unit/test_docling_service.py` covering at minimum:
- `_poll_result` with `success` status on first poll
- `_poll_result` with `failure` status
- `_poll_result` timeout path (`TimeoutError` raised after `timeout` seconds)
- `upload_to_docling_direct_async` HTTP error handling
- `_build_docling_options` with OCR/table-structure toggled on/off

File: src/services/docling_service.py [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+        """
+        Upload a file to Docling Serve asynchronously using direct multipart/form-data upload.
+        """
+        options = self._build_docling_options()
+        
+        # Docling serve async multipart endpoint /v1/convert/file/async
+        # Options are passed as form data
+        data = {k: str(v) if isinstance(v, bool) else v for k, v in options.items() if not isinstance(v, dict)}
+        # picture_description_local needs to be JSON if it's a dict
+        import json
```
Comment (mpawlow) [inline src/services/docling_service.py]: ### (k) [Minor] `import json` Inside Method Body Instead of Module Level

- Note: This is Minor severity. Please feel free to ignore or optionally implement

#### Problem

- `import json` appears at line 97 inside the body of `upload_to_docling_direct_async`, rather than at the top of the module
- This is a Python anti-pattern: `import` inside a function body is re-executed on every call (though cached by the interpreter, it adds overhead and obscures dependencies)
- The stdlib `json` module is already heavily used throughout the codebase at module level

#### Potential Solution

Move `import json` to the top-level imports section of `docling_service.py`:

```python
import base64
import asyncio
import json       # <-- move here
import platform
import os
...
```

File: src/services/docling_service.py (hunk 9 above)
Comment (mpawlow) [inline src/services/docling_service.py]: ### (l) [Minor] Unused Imports `base64` and `os` in `docling_service.py`

- Note: This is Minor severity. Please feel free to ignore or optionally implement

#### Problem

- `import base64` (line 1) and `import os` (line 4) are present in `docling_service.py` but neither module is referenced anywhere in the file
- These are dead imports that increase cognitive overhead and would be flagged by any linter

#### Potential Solution

Remove both unused imports from the file header.

File: src/services/langflow_file_service.py [hunk 11]
```diff
@@ -209,6 +212,7 @@ async def run_ingestion_flow(
             "X-Langflow-Global-Var-SELECTED_EMBEDDING_MODEL": str(embedding_model),
             "X-Langflow-Global-Var-DOCUMENT_ID": str(document_id) if document_id else "",
             "X-Langflow-Global-Var-SOURCE_URL": str(source_url) if source_url else "",
+            "X-Langflow-Global-Var-DOCLING_TASK_ID": str(docling_task_id) if docling_task_id else "",
```
Comment (mpawlow) [inline src/services/langflow_file_service.py]: ## (m) [Minor] `run_url_ingestion_flow` Does Not Pass `DOCLING_TASK_ID` Global Variable Header

- Note: This is Minor severity. Please feel free to ignore or optionally implement

### Problem

- `run_ingestion_flow` adds `X-Langflow-Global-Var-DOCLING_TASK_ID` to headers (line 215 of `langflow_file_service.py`)
- `run_url_ingestion_flow` builds its own `headers` dict but does not include `X-Langflow-Global-Var-DOCLING_TASK_ID`
- If the Langflow flow references `DOCLING_TASK_ID` as a global variable (which it now does, given the env-var registration), URL ingestion may raise an unresolved-variable error in Langflow, or silently fall back to the env-var default `"None"`
- At minimum the omission creates an inconsistency between the two code paths

### Code References

- `src/services/langflow_file_service.py` — lines 330–342 (`run_url_ingestion_flow` headers dict, missing `DOCLING_TASK_ID`)

### Potential Solution

Add the header to `run_url_ingestion_flow` with an empty string default (URL ingestion does not use Docling):

```python
headers = {
    ...
    "X-Langflow-Global-Var-DOCLING_TASK_ID": "",
}
```

### Alternative Solutions

- Consolidate shared header construction into a private `_build_langflow_global_var_headers(...)` helper, so future additions are automatically applied to both code paths.

**Review by mpawlow — CHANGES_REQUESTED** [review]
@lucaseduoli 

## Code Review 1

- See PR comments: (a) to (m)
- Note: **Functional Review 1** still pending

**Comment by mpawlow** [comment]
@coderabbitai full review

File: src/services/langflow_file_service.py [hunk 12]
```diff
@@ -575,54 +615,29 @@ async def upload_and_ingest_file(
...
-                logger.debug(
-                    "[LF] Deleting file after successful ingestion",
-                    extra={"file_id": file_id},
-                )
-                await self.delete_user_file(file_id)
-                delete_result = {"status": "deleted", "file_id": file_id}
-                logger.debug("[LF] File deleted successfully")
-            except Exception as e:
-                delete_error = str(e)
+                await self.docling_service.cancel_task(task_id)
```
Comment (mpawlow) [inline src/services/langflow_file_service.py]: ## (2a) [Major] `DoclingService.cancel_task` Called But Method Does Not Exist

### Problem

- `upload_and_ingest_file` in `langflow_file_service.py` (lines 627–634) attempts to cancel the Docling task when Langflow ingestion fails:
  ```python
  try:
      await self.docling_service.cancel_task(task_id)
  except Exception as cleanup_err:
      logger.warning(
          "[LF] Failed to cancel Docling task after ingestion error",
          task_id=task_id,
          error=str(cleanup_err),
      )
  ```
- `DoclingService` defines no `cancel_task` method — the class exposes only `upload_to_docling_direct_async`, `get_docling_result_async`, `_poll_result`, `convert_file`, and `convert_bytes`
- Every invocation raises `AttributeError: 'DoclingService' object has no attribute 'cancel_task'`
- The broad `except Exception` clause catches the `AttributeError` and logs a misleading warning that implies a remote call failed, when in fact the method was never defined locally
- The original resource-leak problem (orphaned Docling tasks) from issue (d) is therefore still present

### Code References

- `src/services/docling_service.py` — lines 45–204 (entire `DoclingService` class — no `cancel_task` method present)

### Potential Solution

Implement `cancel_task` in `DoclingService`. Docling-serve exposes `DELETE /v1/tasks/{task_id}` for task cancellation:

```python
async def cancel_task(self, task_id: str) -> None:
    """Cancel a previously submitted Docling task."""
    client = self._get_client()
    should_close = client != self.httpx_client
    try:
        if should_close:
            async with client:
                resp = await client.delete(
                    f"{self.docling_url}/v1/tasks/{task_id}"
                )
        else:
            resp = await client.delete(
                f"{self.docling_url}/v1/tasks/{task_id}"
            )
        resp.raise_for_status()
    except Exception as e:
        logger.warning("Docling task cancellation failed", task_id=task_id, error=str(e))
```

### Alternative Solutions

- If Docling-serve does not support task cancellation via API, document the limitation clearly (add a comment explaining why `cancel_task` is a no-op) and log the orphaned task ID at `WARNING` level so operators can manually clean up.
- Accept the resource leak for now and add a background sweeper in `DoclingService` that periodically calls `/v1/tasks` and deletes tasks older than a configurable TTL.

**Review by mpawlow — CHANGES_REQUESTED** [review]
@lucaseduoli 

## Code Review 2

- All previous PR comments from **Code Review 1** addressed  👍 
- See PR comment: (2a) 
   - Related to changes made for (d)

**Comment by mpawlow** [comment]
@lucaseduoli 

## Functional Review 1

- ✅  All sanity tests passed for the main "green thread" scenarios
- ⚠️  Encountered the following UI-related error on chat
   - Not sure if it is related to the current changes or pre-existing

### UI Error

- Ingested Rico's proposal document
   - https://ibm.box.com/s/k2d42c4l60t15jecebispti9clvucxy1
- Chat: "Tell me about the OpenRAG Status Page proposal"
   - BUG: Exception

image

Frontend Logs

```
[browser] Maximum update depth exceeded. This can happen when a component calls setState inside useEffect, but useEffect either doesn't have a dependency array, or one of the dependencies changes on every render.
    at sendMessage (hooks/useChatStreaming.ts:561:19)
    at async handleSSEStream (app/chat/page.tsx:726:5)
    at async handleSendMessage (app/chat/page.tsx:767:7)
  559 |                   thisStreamId === streamIdRef.current
  560 |                 ) {
> 561 |                   setStreamingMessage({
      |                   ^
  562 |                     role: "assistant",
  563 |                     content: currentContent,
  564 |                     functionCalls: (hooks/useChatStreaming.ts:561:19)
```

File: src/services/langflow_file_service.py (hunk 12 above)
Comment (lucaseduoli) [reply src/services/langflow_file_service.py]: docling serve does not support task cancellation via API, removed that part of the code

**Review by mpawlow — APPROVED** [review]
## Code Review 3

- ✅  Approved / LGTM  🚀
